File: 2020/12.rb
DIRS = [[0, 1], [1, 0], [0, -1], [-1, 0]].freeze
ROTATIONS = [
	lambda {|x, y| return [y, -x]},
	lambda {|x, y| return [-x, -y]}, 
	lambda {|x, y| return [-y, x]}].freeze

def rotate(waypoint, dir, mag)
	case mag
	when "90"
		if dir == "R"
			temp = ROTATIONS[0].call(waypoint[0], waypoint[1])
		else
			temp = ROTATIONS[2].call(waypoint[0], waypoint[1])
		end
	when "270"
		if dir == "R"
			temp = ROTATIONS[2].call(waypoint[0], waypoint[1])
		else
			temp = ROTATIONS[0].call(waypoint[0], waypoint[1])
		end
	else
		temp = ROTATIONS[1].call(waypoint[0], waypoint[1])
	end
	return temp
end

lines = File.open("12.txt").read.split("\n")
p2 = false
2.times do |_|
	pos, currDir, waypoint = [0, 0], 1, [10, 1]
	lines.each do |line|
		instruc, mag = line.match(/(\w)(\d+)/).captures
		if "NESW".include? instruc
			dir = DIRS["NESW".index(instruc)]
			mag.to_i.times do |i|
				if !p2
					pos[0] += dir[0]
					pos[1] += dir[1]
				else
					waypoint[0] += dir[0]
					waypoint[1] += dir[1]
				end
			end
		else
			case instruc
			when "F"
				mag.to_i.times do |i|
					if !p2
						pos[0] += DIRS[currDir][0]
						pos[1] += DIRS[currDir][1]
					else
						pos[0] += waypoint[0]
						pos[1] += waypoint[1]
					end
				end
			when "L"
				if !p2
					currDir = (currDir - (mag.to_i/90)) % DIRS.size
				else
					waypoint = rotate(waypoint, "L", mag)
				end
			else
				if !p2
					currDir = (currDir + (mag.to_i/90)) % DIRS.size
				else
					waypoint = rotate(waypoint, "R", mag)
				end
			end
		end
	end
	p2 = !p2
	puts pos.map {|x| x.abs}.sum
end

File: 2020/12.py
import math

lines = open("12.txt").read().split("\n")

loc = [0, 0]
direc = 0
for line in lines:
    instruc, val = line[0], int(line[1:])
    if instruc == "N":
        loc[1] += val
    elif instruc == "S":
        loc[1] -= val
    elif instruc == "E":
        loc[0] += val
    elif instruc == "W":
        loc[0] -= val
    elif instruc == "L":
        direc = (direc + val) % 360
    elif instruc == "R":
        direc = (direc - val) % 360
    elif instruc == "F":
        if direc == 0:
            loc[0] += val
        elif direc == 90:
            loc[1] += val
        elif direc == 180:
            loc[0] -= val
        elif direc == 270:
            loc[1] -= val

# part 1: easy money
print (sum(map(abs, loc)))

loc = [0, 0]
point = [10, 1]
for line in lines:
    instruc, val = line[0], int(line[1:])
    print (line)
    if instruc == "N":
        point[1] += val
    elif instruc == "S":
        point[1] -= val
    elif instruc == "E":
        point[0] += val
    elif instruc == "W":
        point[0] -= val
    elif instruc == "L":
        val = val * math.pi / 180
        point = [(int(math.cos(val)) * (point[0]) - int(math.sin(val)) * (point[1])),
                 (int(math.sin(val)) * (point[0]) + int(math.cos(val)) * (point[1]))]
    elif instruc == "R":
        if val == 180:
            point = [-1 * point[0], -1 * point[1]]
        else:
            val = val * math.pi / 180
            point = [-1 * (int(math.cos(val)) * (point[0]) - int(math.sin(val)) * (point[1])),
                    -1 * (int(math.sin(val)) * (point[0]) + int(math.cos(val)) * (point[1]))]
    elif instruc == "F":
        loc[0] += point[0] * val
        loc[1] += point[1] * val

# part 1: Took a long time to realize that R180 was giving me the wrong result. ended up not
# doing any rotation at all, instead of a true -180
print (sum(map(abs, loc)))